autogpt_platform/backend/backend/data/execution.py

```diff
@@ -10,62 +20,144 @@
     AgentNodeExecution,
     AgentNodeExecutionInputOutput,
 )
-from prisma.types import AgentNodeExecutionUpdateInput, AgentNodeExecutionWhereInput
+from prisma.types import (
+    AgentGraphExecutionWhereInput,
+    AgentNodeExecutionUpdateInput,
+    AgentNodeExecutionWhereInput,
+)
 from pydantic import BaseModel
+from pydantic.fields import Field
 
-from backend.data.block import BlockData, BlockInput, CompletedBlockOutput
-from backend.data.includes import EXECUTION_RESULT_INCLUDE, GRAPH_EXECUTION_INCLUDE
-from backend.data.model import GraphExecutionStats, NodeExecutionStats
-from backend.data.queue import AsyncRedisEventBus, RedisEventBus
 from backend.server.v2.store.exceptions import DatabaseError
-from backend.util import mock, type
+from backend.util import mock
+from backend.util import type as type_utils
 from backend.util.settings import Config
 
+from .block import BlockData, BlockInput, CompletedBlockOutput
+from .db import BaseDbModel
+from .includes import EXECUTION_RESULT_INCLUDE, GRAPH_EXECUTION_INCLUDE
+from .model import GraphExecutionStats, NodeExecutionStats
+from .queue import AsyncRedisEventBus, RedisEventBus
 
-class GraphExecutionEntry(BaseModel):
-    user_id: str
-    graph_exec_id: str
-    graph_id: str
-    graph_version: int
-    start_node_execs: list["NodeExecutionEntry"]
+T = TypeVar("T")
 
+config = Config()
 
-class NodeExecutionEntry(BaseModel):
+
+# -------------------------- Models -------------------------- #
+
+
+ExecutionStatus = AgentExecutionStatus
+
+
+class GraphExecutionMeta(BaseDbModel):
     user_id: str
-    graph_exec_id: str
+    started_at: datetime
+    ended_at: datetime
+    cost: Optional[int] = Field(..., description="Execution cost in credits")
+    duration: float = Field(..., description="Seconds from start to end of run")
+    total_run_time: float = Field(..., description="Seconds of node runtime")
+    status: ExecutionStatus
     graph_id: str
-    node_exec_id: str
-    node_id: str
-    block_id: str
-    data: BlockInput
+    graph_version: int
+    preset_id: Optional[str] = None
 
+    @staticmethod
+    def from_db(_graph_exec: AgentGraphExecution):
+        now = datetime.now(timezone.utc)
+        start_time = _graph_exec.startedAt or _graph_exec.createdAt
+        end_time = _graph_exec.updatedAt or now
+        duration = (end_time - start_time).total_seconds()
+        total_run_time = duration
+
+        try:
+            stats = type_utils.convert(_graph_exec.stats or {}, dict[str, Any])
```

Comment:
There is `GraphExecutionStats` you can `model_validate` the value from this instead of parsing it manually using dict.

You might also want to add `extra = "allow"` in GraphExecutionStats.Config though